### PR TITLE
Handle and detect #13709's unlock regression

### PR DESCRIPTION
**Pull request by rincebrain**
(This is a DRAFT because I want to see the CI run it in a bunch of places, and I'd like people's comments on the approach. I'm sure there are other ways to do this, but this was the simplest one that occurred to me, and it worked.)

In #13709, as in #11294 before it, it turns out that 63a26454 still had the same failure mode as when it was first landed as d1d47691, and fails to unlock certain datasets that formerly worked.

Rather than reverting it again, let's add handling to just throw out the accounting metadata that failed to unlock when that happens, as well as a test with a pre-broken pool image to ensure that we never get bitten by this again.

(CC: @tonyhutter, since 63a26454 is in 2.1.3-2.1.6 as e257bd48...)

### Motivation and Context
#11294/#13709/a number of others listed in 63a26454...

### Description
Just shrug and allow triggering the existing special-case handling of zeroed MAC for useraccounting needing to be regenerated.

(I know I didn't add it to the man page yet, I'll go do that once there's consensus around naming it and whether it needs a tunable at all...I just wanted to include one in case it turns out there are edge cases I didn't notice...)

### How Has This Been Tested?
On Debian 11 and FreeBSD 13.1, with this change, the test case added passes with the tunable enabling the workaround set to 1 and fails with it set to 0, and fails without the code changes at all.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by rincebrain** [comment]
AFAICT the Fedora failure is just "the buildbot runs on shitty instances, timed out, and got killed" and the zloop failure is "zloop fails every time and nobody bothers debugging it for 6 months at a time"?

**Review by behlendorf — APPROVED** [review]
This makes sense.  In fact, it seems to me this is always the right thing in the case of a zeroed mac, we should probably even drop the module option entirely.  I think we're always going to want to tolerate this case for a very long time to maximize compatibility with those previous ZFS releases.  So I'm good with this change if you can just spruce it up!

**Comment by rincebrain** [comment]
The tunable was because I wasn't confident this was correct in every case, and wanted an escape hatch in case I was mistaken.

If we're good without it, so be it. Tunable dropped, cstyle cleaned up, squashed, and pushed.

File: module/zfs/dsl_crypt.c [hunk 1]
```diff
@@ -2671,6 +2671,7 @@ spa_do_crypt_objset_mac_abd(boolean_t generate, spa_t *spa, uint64_t dsobj,
 	objset_phys_t *osp = buf;
 	uint8_t portable_mac[ZIO_OBJSET_MAC_LEN];
 	uint8_t local_mac[ZIO_OBJSET_MAC_LEN];
+	uint8_t zeroed_mac[ZIO_OBJSET_MAC_LEN] = {0};
```
Comment (tonyhutter) [inline module/zfs/dsl_crypt.c:2674]: nit: zeroed_mac should be const

**Review by tonyhutter — APPROVED** [review]

**Comment by tonyhutter** [comment]
@rincebrain do you still want this marked DRAFT, or are you pretty satisfied with it now?

**Comment by rincebrain** [comment]
Good point. Fixed.
